ReactNativeClient/lib/models/Tag.js

```diff
@@ -174,14 +210,121 @@ class Tag extends BaseItem {
+			childItem.title = childName;
+			// Change the child title
+			await Tag.save(childItem, { selectNoteTag: false, createParents: false });
+			// Change descendant titles
+			await Tag.changeDescendantPrefix(childItem, childName);
+		}
+	}
+
+	static async rename(tag, newTitle) {
+		tag.title = newTitle;
+		let oldParentId = tag.parent_id;
+
+		tag = await Tag.save(tag, { fields: ['title', 'parent_id'], userSideValidation: true });
+		await Tag.changeDescendantPrefix(tag, newTitle);
+
+		if (oldParentId !== tag.parent_id) {
+			// If the parent tag has changed, and the ancestor doesn't
+			// have notes attached, then remove it
+			while (oldParentId) {
+				const oldParent = await Tag.load(oldParentId);
+				const oldParentWithCount = await Tag.loadWithCount(oldParentId);
+				if (!oldParentWithCount) {
+					await Tag.delete(oldParentId, { deleteChildren: false });
+				}
+				oldParentId = oldParent.parent_id;
+			}
+		}
+		return tag;
+	}
+
+	static async moveTag(tagId, parentTagId) {
+		const tag = await Tag.load(tagId);
+		if (!tag) return;
+		let tagTitle = await Tag.displayTitle(tag);
+
+		const parentTag = await Tag.load(parentTagId);
+		if (parentTag) {
+			tagTitle = `${parentTag.title}/${tagTitle}`;
+		}
+
+		return await Tag.rename(tag, tagTitle);
+	}
+
 	static async save(o, options = null) {
-		if (options && options.userSideValidation) {
-			if ('title' in o) {
-				o.title = o.title.trim().toLowerCase();
+		if (!options) options = {};
+		if (!('createParents' in options)) options.createParents = true;
+		if (!('selectNoteTag' in options)) options.selectNoteTag = true;
+
+		if ('title' in o) {
+			o.title = o.title.trim().toLowerCase();
+			const existingTagByTitle = await Tag.loadByTitle(o.title);
+
+			if (options && options.userSideValidation && existingTagByTitle && existingTagByTitle.id !== o.id) {
+				throw new Error(_('The tag "%s" already exists. Please choose a different name.', o.title));
+			}
+
+			let parentId = '';
+			let parentTitle = '';
+			// Check if the tag is nested using `/` as separator
+			const separator = '/';
+			const i = o.title.lastIndexOf(separator);
+			if (i !== -1) {
+				parentTitle = o.title.slice(0,i);
+
+				// Try to get the parent tag
+				const parentTag = await Tag.loadByTitle(parentTitle);
+				if (parentTag) parentId = parentTag.id;
+			}
 
-				const existingTag = await Tag.loadByTitle(o.title);
-				if (existingTag && existingTag.id !== o.id) throw new Error(_('The tag "%s" already exists. Please choose a different name.', o.title));
+			if (options && options.createParents && parentTitle && !parentId) {
+				// Create the parent tag if it doesn't exist
+				const parent = {
+					title: parentTitle,
+				};
+				// Do not select the parent note_tags
+				const parentOpts = {};
+				parentOpts.selectNoteTag = false;
+				const parentTag = await Tag.save(parent, parentOpts);
+				parentId = parentTag.id;
 			}
+			// Set parent_id
+			o.parent_id = parentId;
```

Comment:
The `save()` method is going to be used during sync so the data shouldn't be changed at this point as it can create strange inconsistency. It should not load data either as that data may or may not be there during the sync process. In general, this kind of business logic should be done before, when it's needed, and not when saving.

Reply:
Thanks, I do need to look into the synchronizer, and will then get this fixed.

Reply:
Created a `Tag.saveNested()` method, that handles nested tag creation and kept the `save()` method untouched.